src/hotspot/share/oops/constantPool.cpp

```diff
@@ -535,29 +537,40 @@ Klass* ConstantPool::klass_at_impl(const constantPoolHandle& this_cp, int which,
   // Failed to resolve class. We must record the errors so that subsequent attempts
   // to resolve this constant pool entry fail with the same error (JVMS 5.4.3).
   if (HAS_PENDING_EXCEPTION) {
-    if (save_resolution_error) {
-      save_and_throw_exception(this_cp, which, constantTag(JVM_CONSTANT_UnresolvedClass), CHECK_NULL);
-      // If CHECK_NULL above doesn't return the exception, that means that
-      // some other thread has beaten us and has resolved the class.
-      // To preserve old behavior, we return the resolved class.
-      klass = this_cp->resolved_klasses()->at(resolved_klass_index);
-      assert(klass != NULL, "must be resolved if exception was cleared");
-      return klass;
-    } else {
-      return NULL;  // return the pending exception
-    }
+    save_and_throw_exception(this_cp, which, constantTag(JVM_CONSTANT_UnresolvedClass), CHECK_NULL);
+    // If CHECK_NULL above doesn't return the exception, that means that
+    // some other thread has beaten us and has resolved the class.
+    // To preserve old behavior, we return the resolved class.
+    Klass* klass = this_cp->resolved_klasses()->at(resolved_klass_index);
+    assert(klass != NULL, "must be resolved if exception was cleared");
```

Comment:
Existing code but this seems a bizarre thing to do. The only way the two resolving threads can disagree about the resolution is if we have a bad classloader. With a bad classloader it has to justify any argument about what should happen, so it would have seemed preferrable to me to always report the error. It is a race so the outcome is arbitrary to begin with.

Reply:
Yes, you need a non well behaved class loader.  I think the key point is that the *first* outcome is the one that must be saved as the resolution for this constant pool index.  This bug was that if the first was already saved as an error, we were overwriting it.